src/main/kotlin/nikolay/egorov/cli/analysis/lexical/Lexer.kt

```diff
@@ -0,0 +1,101 @@
+package nikolay.egorov.cli.analysis.lexical
+
+import nikolay.egorov.cli.analysis.exceptions.LexException
+
+/**
+ * Base class for lexical analysis phase
+ * @see LexerInterface
+ */
+class Lexer(givenString: String) : LexerInterface {
+
+    private val fedString: String = givenString
+    private val tokens: ArrayList<Lexem> = ArrayList()
+    private var currPos = 0
+    private val stopLexems = setOf('=', '|', '\u0000', '\'', '"', '\t', ' ')
+
+    override fun pushLexem(type: LexemType, lexem: String, positionOfStart: Int) {
+        var actualStart = positionOfStart
+        if (actualStart == -1) {
+            actualStart = currPos
+        }
+        tokens.add(Lexem(type, lexem, actualStart, currPos - 1))
+    }
+
+    override fun lexInput(): ArrayList<Lexem> {
+        val stopLength = fedString.length
+        if (stopLength == 0) {
+            return tokens
+        }
+        var currentChar = fedString[currPos]
+        var currPos_ = Integer(0)
+        while (currPos <= stopLength - 1) {
+            val posWas = currPos
+            when (currentChar) {
+                in "\'\"" -> lexQuotedWord(currentChar)
+                in "=|" -> lexOperation(currentChar)
+                in " \t" -> currPos++
+                else -> lexWord(currentChar)
+            }
+            if (posWas == currPos) {
+                currPos++
+            }
+            currentChar = lookAt()
+        }
+
+        return tokens
+    }
+
+    override fun lexWord(startsWith: Char) {
+        val buf = StringBuilder()
+        val startOnPosition = currPos
+        var currentChar = startsWith
+
+        while (!stopLexems.contains(currentChar)) {
+            buf.append(currentChar)
+            currPos++
+            currentChar = lookAt()
+        }
+
+        pushLexem(LexemType.WORD, buf.toString(), startOnPosition)
+    }
+
+    override fun lexQuotedWord(startsWith: Char) {
+        val buf = StringBuilder()
+        val startOnPosition = currPos++
+        var currentChar = lookAt()
+
+        while (true) {
+            if (currentChar == startsWith) break
+            if (currentChar == '\u0000')
+                throw LexException("Unexpected end of the word", currPos)
+            buf.append(currentChar)
+            currPos++
+            currentChar = lookAt()
+        }
```

Comment:
Немного странное решение про `lookAt`. С ней сложнее понимать, из-за того, что она возвращает магический '\u0000'. Во-первых это магическая константа, во-вторых кажется правильнее проверять до её вызова, вышли ли за границу строки, а внутри lookAt уже бросать исключение, если за границей

Кажется можно переписать так

```suggestion
        while (currPos < givenString.length) {
            val currentChar = givenString[currPos]
            if (currentChar == startsWith) break
            buf.append(currentChar)
            currPos++
        }
```